From d3b446c8280a9a91116a3bec5dca43a9a55c2b5b Mon Sep 17 00:00:00 2001
From: nikhilad <nikhilad@ami.com>
Date: Thu, 18 Sep 2025 10:21:14 +0530
Subject: [PATCH] Optimizing the Thermal OEM Conditions code

Removed unneccesary io connections
and implemented optimized code to prevent repetitive function calls.Wq

Signed-off-by: nikhilad <nikhilad@ami.com>
---
 pid/actioncontroller.hpp | 35 ++++-------------------------------
 pid/zone.hpp             | 28 +++++++++++++++++++---------
 2 files changed, 23 insertions(+), 40 deletions(-)

diff --git a/pid/actioncontroller.hpp b/pid/actioncontroller.hpp
index e947b44..9abe56d 100644
--- a/pid/actioncontroller.hpp
+++ b/pid/actioncontroller.hpp
@@ -152,33 +152,6 @@ inline bool getPowerStatus(std::shared_ptr<sdbusplus::asio::connection> conn)
     return pwrGood;
 }
 
-inline bool powerOn(std::shared_ptr<sdbusplus::asio::connection> conn)
-{
-    bool pwrGood = false;
-    std::string pwrStatus;
-    value variant;
-    try
-    {
-        auto method = conn->new_method_call(pwrService, pwrHostStateObjPath,
-                                            PROP_INTF, METHOD_GET);
-        method.append(pwrHostStateIface, "CurrentHostState");
-        auto reply = conn->call(method);
-        reply.read(variant);
-        pwrStatus = std::get<std::string>(variant);
-    }
-    catch (sdbusplus::exception_t& e)
-    {
-        phosphor::logging::log<phosphor::logging::level::ERR>(
-            "Failed to get getPowerStatus Value",
-            phosphor::logging::entry("EXCEPTION=%s", e.what()));
-        return pwrGood;
-    }
-    if (pwrStatus == "xyz.openbmc_project.State.Host.HostState.Off")
-    {
-        pwrGood = true;
-    }
-    return pwrGood;
-}
 inline void poweroff(std::shared_ptr<sdbusplus::asio::connection> conn)
 {
     std::cerr << " poweroff \n";
@@ -198,10 +171,10 @@ inline void poweroff(std::shared_ptr<sdbusplus::asio::connection> conn)
 inline std::string getService(const std::string& intf, const std::string& path)
 {
     auto bus = bus::new_default_system();
-    auto mapper = bus.new_method_call("xyz.openbmc_project.ObjectMapper",
-                                      "/xyz/openbmc_project/object_mapper",
-                                      "xyz.openbmc_project.ObjectMapper",
-                                      "GetObject");
+    auto mapper =
+        bus.new_method_call("xyz.openbmc_project.ObjectMapper",
+                            "/xyz/openbmc_project/object_mapper",
+                            "xyz.openbmc_project.ObjectMapper", "GetObject");
 
     mapper.append(path);
     mapper.append(std::vector<std::string>({intf}));
diff --git a/pid/zone.hpp b/pid/zone.hpp
index 12d102c..b560385 100644
--- a/pid/zone.hpp
+++ b/pid/zone.hpp
@@ -145,6 +145,16 @@ class DbusPidZone : public ZoneInterface, public ModeObject, public ActionOem
         uint8_t readFailureCnt = 0;
         boost::asio::io_context io;
         auto conn = std::make_shared<sdbusplus::asio::connection>(io);
+        bool powerStatus = getPowerStatus(conn);
+
+        auto isFanSensor = [&](const std::string& name) {
+            return (name.find(aspeedFan) != std::string::npos) ||
+                   (name.find(psuFan) != std::string::npos);
+        };
+
+        auto isTempSensor = [&](const std::string& name) {
+            return (name.find(tempSensor) != std::string::npos);
+        };
 
         for (const auto& sensorInput : sensorInputs)
         {
@@ -155,21 +165,21 @@ class DbusPidZone : public ZoneInterface, public ModeObject, public ActionOem
             std::string sensorName = sensor->getName();
             if (fanSensorLogging == true)
             {
-                if (((sensorName.find(aspeedFan)) || (sensorName.find(psuFan))))
+                if (isFanSensor(sensorName))
                 {
-                    static boost::asio::io_context io;
-                    static boost::asio::steady_timer PowerDelay(io);
                     double setpoint = 0;
 
-                    if (powerOn(conn))
-                    {
-                        tryRestartControlLoops();
-                    }
-                    if (getPowerStatus(conn))
+                    if (powerStatus)
                     {
+                        syslog(LOG_INFO, "Nik-swampd: Power is ON \n ");
                         setpoint = processFanAction(r.value, sensorName,
                                                     &readFailureCnt);
                     }
+                    else
+                    {
+                        tryRestartControlLoops();
+                    }
+
                     if (setpoint > 0)
                     {
                         DbusPidZone::addSetPoint(setpoint, sensorName);
@@ -178,7 +188,7 @@ class DbusPidZone : public ZoneInterface, public ModeObject, public ActionOem
             }
             else
             {
-                if ((sensorName.find(tempSensor)))
+                if (isTempSensor(sensorName))
                 {
                     double setpoint = 0;
 
-- 
2.34.1

